src/preamble.js

```diff
@@ -129,35 +129,28 @@ function ccall(ident, returnType, argTypes, args, opts) {
     }
   }
   var ret = func.apply(null, cArgs);
+  function onDone(ret) {
+    if (stack !== 0) stackRestore(stack);
+    return convertReturnValue(ret);
+  }
 #if ASYNCIFY
   var asyncMode = opts && opts.async;
-  var runningAsync = typeof Asyncify === 'object' && Asyncify.currData;
-  var prevRunningAsync = typeof Asyncify === 'object' && Asyncify.asyncFinalizers.length > 0;
-#if ASSERTIONS
-  assert(!asyncMode || !prevRunningAsync, 'Cannot have multiple async ccalls in flight at once');
-#endif
   // Check if we started an async operation just now.
-  if (runningAsync && !prevRunningAsync) {
+  if (Asyncify.currData) {
```

Comment:
I think this introduces a bug so that during an async function, we can no longer ccall a benign sync function like a thin wrapper around malloc, increment a static counter, etc. This makes it very troublesome for a complex loading task to complete.

Reply:
I believe this part of the diff is inconsistent with the guidance here: https://github.com/emscripten-core/emscripten/pull/14664#discussion_r672586792

Reply:
I think you're right @justjake 

@RReverser was that intentional?

Reply:
I think it was (didn't we settle on direct `Module._foo()` calls as enough for a rare escape hatch?) but I don't remember for sure, and I'm not prepared to dig in right now.

Reply:
Ok, no worries, I'll look into it.

Reply:
Thanks for looking @kripken - it's awesome you're responsive to a comment on a random commit ❤️ 
I got distracted and forgot to do a full repro/bug report, but this issue is affecting me; currently it's blocking my work here: https://github.com/justjake/quickjs-emscripten/pull/54

Reply:
Would more of a write-up be helpful?

Reply:
Maybe, but first, can you test if #16370 fixes things for you? (Skimming that link, I think/hope it should.)